processing/src/main/java/org/apache/druid/query/expression/IPv4AddressMatchExprMacro.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.expression;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.net.util.SubnetUtils;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.math.expr.Expr;
+import org.apache.druid.math.expr.ExprEval;
+import org.apache.druid.math.expr.ExprMacroTable;
+import org.apache.druid.math.expr.ExprType;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * <pre>
+ * Implements an expression that checks if an IPv4 address belongs to a particular subnet.
+ *
+ * Expression signatures:
+ * - boolean ipv4address_match(string address, string subnet)
+ * - boolean ipv4address_match(string address, string subnet, boolean inclusive)
+ * - boolean ipv4address_match(long address, string subnet)
+ * - boolean ipv4address_match(long address, string subnet, boolean inclusive)
+ *
+ * Valid "address" argument formats are:
+ * - unsigned int long (e.g., 3232235521)
+ * - unsigned int string (e.g., "3232235521")
+ * - IPv4 address dotted-decimal notation string (e.g., "198.168.0.1")
+ * - IPv6 IPv4-mapped address string (e.g., "::ffff:192.168.0.1")
+ *
+ * The argument format for the "subnet" argument should be a literal in CIDR notation
+ * (e.g., "198.168.0.0/16").
+ *
+ * An optional "inclusive" argument should be a boolean literal (e.g., 1, 0, "true", or "false") that
+ * indicates whether the network and the broadcast addresses should be considered part of the
+ * subnet. When this argument is absent, its value defaults to "false".
+ *
+ * The overloaded signature allows applying the expression to a dimension with mixed string and long
+ * representations of IPv4 addresses.
+ * </pre>
+ *
+ * @see IPv4AddressParseExprMacro
+ * @see IPv4AddressStringifyExprMacro
+ */
+public class IPv4AddressMatchExprMacro implements ExprMacroTable.ExprMacro
```

Comment:
I made `match` an `ExprMacro` so that it could pre-compute the subnet matcher at parse time. `parse` and `stringify` don't do any precomputation, but since I wanted to colocate related expressions, I made both of those `ExprMacros` too.

Reply:
Is`SubnetInfo` expensive to compute? Also, is it best that it only takes literal arguments and not be able to write an expression to compute the subnet argument?

There is probably a sort of janky way that stuff could be pre-computed in `Function` involving using `validateArguments` since that is called at parse time, but maybe it would be nice to plumb a more official way to do such things in the future.

Side note, please don't mark comments as resolved, leave that for the reviewer who made the comment to do.

Reply:
Building `SubnetInfo` requires validating if the subnet argument has a valid (CIDR-notation) format, which only needs to be done once. What kinds of expressions are you thinking of for the subnet argument?

I've undone all the comment threads that I marked as resolved (was using that to keep track of the ones I've addressed in my branch).

Reply:
> There is probably a sort of janky way that stuff could be pre-computed in Function involving using validateArguments since that is called at parse time, but maybe it would be nice to plumb a more official way to do such things in the future.

The purpose of `ExprMacro` is to provide a way to precompute certain things before doing the row-by-row evaluation, so IMO it does make more sense to use ExprMacro (which is designed for this) rather than hook into parse-time Function methods.

> Any reason these functions defined via ExprMacro in druid-processing instead of in Function.java in druid-core?

It's probably an accident that Function and ExprMacro are in different packages. I added ExprMacro in #4365 and I don't recall having a good reason for having them be in different places. Fwiw, I don't have a strong separation my mind between the responsibility of processing and core, anyway, largely agreeing with this comment: https://github.com/apache/incubator-druid/issues/4312#issuecomment-356127914

Reply:
>The purpose of ExprMacro is to provide a way to precompute certain things before doing the row-by-row evaluation, so IMO it does make more sense to use ExprMacro (which is designed for this) rather than hook into parse-time Function methods.

Thanks for the explanation, :+1: I had stored away in my head for whatever reason that the main purpose of `ExprMacro` was so extensions could define expressions. I wonder if there is a way we could cut down the amount of code/test that needs to be added for `ExprMacro` that live in main Druid, because it does seem to take a lot more to add an `ExprMacro` compared to a `Function`, but I don't have any good ideas at the moment, nor am I sure that allowing `Parser.getFunction` to also take the list of args and allow initializing a `Function` when creating the `FunctionExpr` would be the appropriate.